web3/providers/persistent/subscription_manager.py

```diff
@@ -0,0 +1,143 @@
+from web3.utils.subscriptions import (
+    EthSubscription,
+)
+
+if TYPE_CHECKING:
+    from web3 import AsyncWeb3  # noqa: F401
+    from web3.providers.persistent import PersistentConnectionProvider  # noqa: F401
+
+
+class SubscriptionManager:
+    logger: logging.Logger = logging.getLogger(
+        "web3.providers.persistent.subscription_manager"
+    )
+    total_handler_calls: int = 0
+
+    def __init__(self, w3: "AsyncWeb3") -> None:
+        self._w3 = w3
+        self._provider = cast("PersistentConnectionProvider", w3.provider)
+        self.subscriptions: List[EthSubscription[Any]] = []
+        self._subscriptions_by_id: Dict[HexStr, EthSubscription[Any]] = {}
+        self._subscriptions_by_label: Dict[str, EthSubscription[Any]] = {}
+
+    def get_by_id(self, sx_id: HexStr) -> EthSubscription[Any]:
+        return self._subscriptions_by_id.get(sx_id)
+
+    def get_by_label(self, label: str) -> EthSubscription[Any]:
+        return self._subscriptions_by_label.get(label)
+
+    @overload
+    async def subscribe(self, subscriptions: EthSubscription[Any]) -> HexStr:
+        ...
+
+    @overload
+    async def subscribe(
+        self, subscriptions: Sequence[EthSubscription[Any]]
+    ) -> List[HexStr]:
+        ...
+
+    async def subscribe(
+        self, subscriptions: Union[EthSubscription[Any], Sequence[EthSubscription[Any]]]
+    ) -> Union[HexStr, List[HexStr]]:
+        if isinstance(subscriptions, EthSubscription):
+            if subscriptions.label in self._subscriptions_by_label:
+                raise Web3ValueError(
+                    "Subscription label already exists. Subscriptions must have "
+                    f"unique labels.\n    label: {subscriptions.label}"
+                )
+
+            subscriptions.manager = self
+            sx_id = await self._w3.eth._subscribe(*subscriptions.subscription_params)
+            subscriptions._id = sx_id
+            self._subscriptions_by_label[subscriptions.label] = subscriptions
+            self._subscriptions_by_id[subscriptions.id] = subscriptions
+            self.subscriptions.append(subscriptions)
+            self.logger.info(
+                "Successfully subscribed to subscription:\n    "
+                f"label: {subscriptions.label}\n    id: {sx_id}"
+            )
+            return sx_id
+        elif isinstance(subscriptions, Sequence):
+            if len(subscriptions) == 0:
+                raise Web3ValueError("No subscriptions provided.")
+
+            sx_ids: List[HexStr] = []
+            for sx in subscriptions:
+                await self.subscribe(sx)
+            return sx_ids
+        raise Web3TypeError("Expected a Subscription or a sequence of Subscriptions.")
+
+    async def unsubscribe(self, subscription: EthSubscription[Any]) -> bool:
+        """
+        Used to unsubscribe from a subscription that is being managed by the
+        subscription manager.
+
+        :param subscription: The subscription to unsubscribe from.
```

Comment:
Might include the type in the docs as well with `:type subscription: EthSubscription[Any]`